forte/orbital-helpers/embedding_density.cc

```diff
@@ -0,0 +1,253 @@
+    } else if (options_->get_str("FRAGMENT_DENSITY") == "CASSCF") {
+        CASSCF cas_1(state_weights_map_, scf_info_, options_, mo_space_info_active, ints_);
+        cas_1.compute_energy();
+        psi::outfile->Printf(
+            "\n CASSCF(A) density will be written into the fragment (A) densiy blocks");
+        ref_rdms = cas_1.ref_rdms();
+    }
+
+    // Build RDMs with mo_space_info size, fill the blocks
+    size_t na = mo_space_info_->size("ACTIVE");
+    size_t na_in = mo_space_info_active->size("ACTIVE");
+
+    // Outer-layer mospace
+    auto doccpi = scf_info_->doccpi();
+    auto rdoccpi = mo_space_info_->dimension("RESTRICTED_DOCC");
+    auto actvpi = mo_space_info_->dimension("ACTIVE");
+    auto actvopi = doccpi - rdoccpi - mo_space_info_->dimension("FROZEN_DOCC");
+
+    // Inner-layer mospace
+    auto rdoccpi_in = mo_space_info_active->dimension("RESTRICTED_DOCC");
+    auto actvpi_in = mo_space_info_active->dimension("ACTIVE");
+    // auto actvopi_in = doccpi - rdoccpi_in - mo_space_info_active->dimension("FROZEN_DOCC");
+
+    std::vector<size_t> mos_oa;
+    for (int i = 0; i < rdoccpi_in[0] + actvpi_in[0]; ++i) {
+        mos_oa.push_back(i);
+    }
+
+    std::vector<size_t> mos_actv_in;
+    for (int i = 0; i < actvpi_in[0]; ++i) {
+        mos_actv_in.push_back(i);
+    }
+
+    // 1-RDM
+    auto D1a = ambit::Tensor::build(ambit::CoreTensor, "D1a", std::vector<size_t>(2, na));
+    auto D1b = ambit::Tensor::build(ambit::CoreTensor, "D1b", std::vector<size_t>(2, na));
+
+    for (auto i : mos_oa) {
+        for (auto j : mos_oa) {
+            if (i >= rdoccpi_in[0] && j >= rdoccpi_in[0]) {
+                size_t ip = i - rdoccpi_in[0];
+                size_t jp = j - rdoccpi_in[0];
+                D1a.data()[i * na + j] = ref_rdms.g1a().data()[ip * na_in + jp];
+                D1b.data()[i * na + j] = ref_rdms.g1b().data()[ip * na_in + jp];
+            } else if (i == j) {
+                D1a.data()[i * na + j] = 1.0;
+                D1b.data()[i * na + j] = 1.0;
+            } else {
+                D1a.data()[i * na + j] = 0.0;
+                D1b.data()[i * na + j] = 0.0;
+            }
+        }
+    }
+
+    // 2-RDM
+    auto D2aa = ambit::Tensor::build(ambit::CoreTensor, "D2aa", std::vector<size_t>(4, na));
+    auto D2ab = ambit::Tensor::build(ambit::CoreTensor, "D2ab", std::vector<size_t>(4, na));
+    auto D2bb = ambit::Tensor::build(ambit::CoreTensor, "D2bb", std::vector<size_t>(4, na));
+
+    D2aa("pqrs") += D1a("pr") * D1a("qs");
+    D2aa("pqrs") -= D1a("ps") * D1a("qr");
+
+    D2bb("pqrs") += D1b("pr") * D1b("qs");
+    D2bb("pqrs") -= D1b("ps") * D1b("qr");
+
+    D2ab("pqrs") = D1a("pr") * D1b("qs");
+
+    size_t v = rdoccpi_in[0];
+
+    for (auto p : mos_actv_in) {
+        for (auto q : mos_actv_in) {
+            for (auto r : mos_actv_in) {
+                for (auto s : mos_actv_in) {
+                    D2aa.data()[index_exp(na, p + v, q + v, r + v, s + v)] =
```

Comment:
Maybe we should not call `.data()` inside the loop. We could say `auto& D2aa_data = D2aa.data();`, etc. outside the loop.